tokenizer_ts/src/bytePairEncode.ts

```diff
@@ -77,14 +79,25 @@ export function bytePairEncode(
     return [ranks.get(mergingBytes)!];
   }
 
+  let minRank = maxRank;
+  let minIndex = -1;
+
   const byteIndicesAndRanks: [number, number][] = [];
```

Comment:
I spent a bit of time investigating whether a single Uint32Array with "marked" removals would be better here, but the performance was about the same for much more complexity. I attribute this to the fact that we had to spend time iterating back and forth to 'skip over' marked removals in that case.

But we may be able to do something similar: have a single Uint32Array and have byteIndicesAndRanks be subarrays of that single memory range. An investigation for tomorrow.

Reply:
I'm actually still a bit unsure about what this code does, but it looks sane!

Reply:
The goal of this function is to compress the input string to the byte-pair "dictionary", which maps sequences of bytes/charater to their 'dictionary number'.

This function creates a list of bytes in the string and their "rank". The rank is a score indicating how 'good' the possible byte sequence at each position it. At each iteration it removes the best-ranked byte and replaces it with its dictionary number, and it then re-ranks the remaining sequences until there the string is fully encoded.